agent/ae/ae.go

```diff
@@ -296,34 +268,66 @@ func (s *StateSyncer) syncChangesEventFn() event {
 // resetNextFullSyncCh resets nextFullSyncCh and sets it to interval+stagger.
 // Call this function everytime a full sync is performed.
 func (s *StateSyncer) resetNextFullSyncCh() {
-	if s.stagger != nil {
-		s.nextFullSyncCh = time.After(s.Interval + s.stagger(s.Interval))
-	} else {
-		s.nextFullSyncCh = time.After(s.Interval)
-	}
+	s.nextFullSyncCh = time.After(s.Interval + s.Delayer.Jitter(s.Interval))
 }
 
-// stubbed out for testing
+// shim for testing
 var libRandomStagger = lib.RandomStagger
 
-// staggerFn returns a random duration which depends on the cluster size
+func NewClusterSizeDelayer(size func() int) Delayer {
+	return delayer{fn: size}
+}
+
+type delayer struct {
+	fn func() int
+}
+
+// Jitter returns a random duration which depends on the cluster size
 // and a random factor which should provide some timely distribution of
-// cluster wide events. This function should not be called directly
-// but through s.stagger to allow mocking for testing.
-func (s *StateSyncer) staggerFn(d time.Duration) time.Duration {
-	f := scaleFactor(s.ClusterSize())
-	return libRandomStagger(time.Duration(f) * d)
+// cluster wide events.
+func (d delayer) Jitter(delay time.Duration) time.Duration {
+	return libRandomStagger(time.Duration(scaleFactor(d.fn())) * delay)
+}
+
+// scaleThreshold is the number of nodes after which regular sync runs are
+// spread out farther apart. The value should be a power of 2 since the
+// scale function uses log2.
+//
+// When set to 128 nodes the delay between regular runs is doubled when the
+// cluster is larger than 128 nodes. It doubles again when it passes 256
+// nodes, and again at 512 nodes and so forth. At 8192 nodes, the delay
+// factor is 8.
+//
+// If you update this, you may need to adjust the tuning of
+// CoordinateUpdatePeriod and CoordinateUpdateMaxBatchSize.
+const scaleThreshold = 128
+
+// scaleFactor returns a factor by which the next sync run should be delayed to
+// avoid saturation of the cluster. The larger the cluster grows the farther
+// the sync runs should be spread apart.
+//
+// The current implementation uses a log2 scale which doubles the delay between
+// runs every time the cluster doubles in size.
+func scaleFactor(nodes int) int {
+	if nodes <= scaleThreshold {
+		return 1.0
+	}
+	return int(math.Ceil(math.Log2(float64(nodes))-math.Log2(float64(scaleThreshold))) + 1.0)
 }
 
 // Pause temporarily disables sync runs.
 func (s *StateSyncer) Pause() {
 	s.pauseLock.Lock()
 	s.paused++
+
+	if s.chPaused == nil {
+		s.chPaused = make(chan struct{})
```

Comment:
Good question! I've also been looking for ways to reduce the number of fields we need for some of this coordination, but in this case I think they may all be necessary.

This pattern of closing a channel (instead of sending a `bool` on a channel) is common in Consul and some of the hashicorp libraries. There is an important difference to sending a value on a channel and closing it.  When a value is sent, only a single consumer of the channel will see it, but when a channel is closed, all readers of the channel will see this notification. I think closing of channels is used all over for this broadcast behaviour. It removes the need to use a separate channel for every consumer.

I believe that is the case here as well. We could have many goroutines blocking on a read from this channel, so we need to use close instead of sending a bool.

Reply:
yes I understand, haven't thought about the multiple goroutine scenario, in that case can we convert the channel here for https://golang.org/pkg/sync/#Cond.Signal. In understand it's a more low level api but in that case I think this is exactly the intended behaviour, what you think?

Reply:
We will need to test for the `paused` state before waiting for the signal though

Reply:
I think you're right, `sync.Cond` seems like a really good fit here. I have never really used `sync.Cond`, so I'd probably want to make that change in a separate follow up PR, to draw more attention to that change. I think the changes here are mostly only moving things around.